File: backend/app/schemas/conversation.py
"""Pydantic schemas for the ``conversations`` resource."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Request ─────────────────────────────────────────────────────

class ConversationCreate(BaseModel):
    """Create a new conversation for the current user.

    All fields are optional — the API will auto-generate defaults.
    A ``title`` may be provided by the client for display purposes,
    otherwise the LLM will generate one from the first user message.
    """

    chart_context_id: UUID | None = None
    synastry_context_id: UUID | None = None
    title: str | None = None
    model_version: str | None = None


class ConversationUpdate(BaseModel):
    """Partial update of a conversation.

    Supports renaming the conversation and changing its status.
    """

    title: str | None = None
    status: str | None = Field(None, pattern="^(active|archived)$")


# ── Response ────────────────────────────────────────────────────

class ConversationResponse(BaseModel):
    """Full conversation returned to the client."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    chart_context_id: UUID | None = None
    synastry_context_id: UUID | None = None
    title: str | None = None
    status: str = "active"
    model_version: str | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/conversation.ts
/* ── Conversation & Message types ────────────────────── */

export type MessageRole = 'system' | 'user' | 'assistant' | 'tool';
export type ConversationStatus = 'active' | 'archived';

export interface Message {
  id: number;
  conversation_id: string;
  role: MessageRole;
  content: string;
  tool_call_id?: string;
  tool_name?: string;
  payload?: Record<string, unknown>;
  created_at: string;
}

export interface Conversation {
  id: string;
  user_id: string;
  chart_context_id?: string;
  synastry_context_id?: string;
  title?: string;
  status: ConversationStatus;
  model_version?: string;
  created_at: string;
  updated_at: string;
}

export interface SendMessageResponse {
  user_message: Message;
  assistant_message: Message;
}
